bundles/core/org.eclipse.smarthome.core.thing.xml/src/main/java/org/eclipse/smarthome/core/thing/xml/internal/XmlThingTypeProvider.java

```diff
@@ -42,11 +43,62 @@
  * disappears, any registered {@link ThingType} objects associated with that module are released.
  *
  * @author Michael Grammling - Initial Contribution
- * @author Dennis Nobel - Added locale support
+ * @author Dennis Nobel - Added locale support, Added cache for localized thing types
  * @author Ivan Iliev - Added support for system wide channel types
  */
 public class XmlThingTypeProvider implements ThingTypeProvider {
 
+    private class LocalizedThingTypeKey {
+        public ThingTypeUID uid;
+        public String locale;
+
+        public LocalizedThingTypeKey(ThingTypeUID uid, String locale) {
+            this.uid = uid;
+            this.locale = locale;
+        }
+
+        @Override
+        public int hashCode() {
+            final int prime = 31;
+            int result = 1;
+            result = prime * result + getOuterType().hashCode();
+            result = prime * result + ((locale == null) ? 0 : locale.hashCode());
+            result = prime * result + ((uid == null) ? 0 : uid.hashCode());
+            return result;
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            if (this == obj)
+                return true;
+            if (obj == null)
+                return false;
+            if (getClass() != obj.getClass())
+                return false;
+            LocalizedThingTypeKey other = (LocalizedThingTypeKey) obj;
+            if (!getOuterType().equals(other.getOuterType()))
+                return false;
+            if (locale == null) {
+                if (other.locale != null)
+                    return false;
+            } else if (!locale.equals(other.locale))
+                return false;
+            if (uid == null) {
+                if (other.uid != null)
+                    return false;
+            } else if (!uid.equals(other.uid))
+                return false;
+            return true;
+        }
+
+        private XmlThingTypeProvider getOuterType() {
+            return XmlThingTypeProvider.this;
+        }
+
+    }
+
+    private Map<LocalizedThingTypeKey, ThingType> localizedThingTypeCache = new HashMap<>();
```

Comment:
There are mainly 3 map operations: put, get and iterate with remove
* put and get are not critical
* the iteration over the entryset is also threadsafe, even if someone put something into the map while iterating through it (see http://stackoverflow.com/questions/3768554/is-iterating-concurrenthashmap-values-thread-safe)

So I don´t see a concurrency problem. It might be that one entry stays in the cache, if getThingType is called at the same time while a binding is removed. To avoid this I would have to synchronize getThingType and getThingTypes against the "add" and "remove" methods.